include/envoy/stats/symbol_table.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include <vector>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Stats {
+/**
+ * Interface for shortening and retrieving stat names.
+ * Guarantees that x = decode(encode(x)) for any x.
+ *
+ * Even though the symbol table does manual reference counting, curr_counter_ is monotonically
+ * increasing. So encoding "foo", freeing the sole stat containing "foo", and then re-encoding "foo"
+ * will produce a different symbol each time.
+ */
+class SymbolTable {
+public:
+  typedef uint32_t Symbol;
+
+  virtual ~SymbolTable() {}
+
+  /**
+   * Encodes a stat name into a vector of Symbols. Expects the name to be period-delimited.
+   *
+   * @param name the stat name to encode.
+   * @return std::vector<Symbol> the encoded stat name.
+   */
+  virtual std::vector<Symbol> encode(const std::string& name) PURE;
```

Comment:
Why is the fact that this is a vector of symbols made public to the world? Can we use an opaque type here to support multiple implementations (and later performance tuning)?

Reply:
This makes sense -- I think the right way to do this is to 
```
typedef uint32_t Symbol;
typedef std::vector<Symbol> StatName;
```
inside the abstract interface, and then pass SymbolTable::StatName everywhere.

Reply:
Oops, with `using` as above.

Reply:
If you want to keep the fact that the rep is a vector<Symbol> an implementation detail, it should not go into the interface.

I think you want something like:

```
class SymbolizedString {
 public:
  SymbolizedString() {}
  virtual ~SymbolizedString();
  virtual std::string() toString() const PURE;
};

class SymbolizedStringImpl : public SymbolizedString {
 public:
  ~SymbolizedStringImpl() override {
      ... decrement ref count on all symbols via SymbolTable_;
  }
  std::string() toString() const override;
 private:
  std::vector<Symbol> symbols_;
  SymbolTable& symbol_table_;
};
```

Reply:
Done. We pass back a std::unique_ptr<StatName> for this which has free() inside its destructor.